[visionOS] Allow image updates when fullscreen is delegated to Quick Look

yoelhawa opened the pull request:
#### b7c11fccb9fbe96df0a6725fe9c6ea34891cfee5
<pre>
[visionOS] Allow image updates when fullscreen is delegated to Quick Look
<a href="https://bugs.webkit.org/show_bug.cgi?id=281592">https://bugs.webkit.org/show_bug.cgi?id=281592</a>
&lt;<a href="https://rdar.apple.com/136326912">rdar://136326912</a>&gt;

Reviewed by Andy Estes.

When an image source changes while in fullscreen with Quick look,
send that new image source over to Quick Look to update the image
being displayed.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::notifyFinished):
Listen for new image sources to finish loading.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::updateImageSource):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::updateImageSource):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::updateImageSource):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController updateImageSource]):
* Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::getImageMediaDetails):
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::updateImageSource):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::updateImageSource):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
Plumb the new image data to the UI Process if we currently have an
image open in fullscreen via Quick Look.

* Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift:
(item):
(windowOpenedContinuation):
(presentWindow):
(updateImage(_:)):
Wait until the Quick Look window is ready if it isn&apos;t already, then
update the Quick Look scene with the new image.

Canonical link: <a href="https://commits.webkit.org/285443@main">https://commits.webkit.org/285443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9b212363cfe925bcd43edfcb6020e5c79df1af9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72671 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/52096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25469 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/23904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23705 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/23904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75738 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/47126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62554 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/43776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/20021 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/22234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20380 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16917 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/78533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16965 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62563 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16045 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/13186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6827 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47894 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/48961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/50256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/48706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
```diff
@@ -136,6 +140,7 @@ class WebFullScreenManager final : public WebCore::EventListener {
 #endif // ENABLE(VIDEO)
 
 #if ENABLE(QUICKLOOK_FULLSCREEN)
+    FullScreenMediaDetails getImageMediaDetails(CheckedPtr<WebCore::RenderImage>, bool);
```
You should be able to get away with a forward decl for `WebCore::RenderImage`

aprotyas commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
```diff
@@ -136,6 +140,7 @@ class WebFullScreenManager final : public WebCore::EventListener {
 #endif // ENABLE(VIDEO)
 
 #if ENABLE(QUICKLOOK_FULLSCREEN)
+    FullScreenMediaDetails getImageMediaDetails(CheckedPtr<WebCore::RenderImage>, bool);
```
Nit: thoughts on making this an enumeration? Makes call sites a little easier to read.

```
enum class IsUpdating : bool { No, Yes };
FullScreenMediaDetails getImageMediaDetails(CheckedPtr<WebCore::RenderImage>, IsUpdating);
```

yoelhawa replied:
Seems fair!

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1115,6 +1115,20 @@ - (void)_enterFullScreen:(CGSize)mediaDimensions windowScene:(UIWindowScene *)wi
+    if (!manager)
+        return;
+
+    manager->prepareQuickLookImageURL([strongSelf = retainPtr(self), self, window = retainPtr([webView window]), logIdentifier = OBJC_LOGIDENTIFIER] (URL&& url) mutable {
```
Can you remove the space between the capture list and the arg list? https://webkit.org/code-style-guidelines/#spacing-lambda-paren

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1115,6 +1115,20 @@ - (void)_enterFullScreen:(CGSize)mediaDimensions windowScene:(UIWindowScene *)wi
+        return;
+
+    manager->prepareQuickLookImageURL([strongSelf = retainPtr(self), self, window = retainPtr([webView window]), logIdentifier = OBJC_LOGIDENTIFIER] (URL&& url) mutable {
+        [_previewWindowController updateImage:url];
```
You might as well `[_previewWindowController updateImage:WTFMove(url)]`

aprotyas reviewed (commented):
LGTM with a couple of small questions!

I'll defer the r+ to a reviewer who's more experienced with fullscreen presentation, though.

etiennesegonzac commented on Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift:
```diff
@@ -79,6 +88,28 @@ public final class PreviewWindowController: NSObject {
+                    }
+                    try await withCheckedThrowingContinuation { continuation in
+                        self.windowOpenedContinuation = continuation
+                    }
```
maybe we should catch `newUpdateQueued` separately since we use it to cancel the previous update but it shouldn't, I think, be logged as an error

yoelhawa replied:
Good catch. Changed that case to a debug log

aestes commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.cpp:
```diff
@@ -61,6 +61,13 @@ void InjectedBundlePageFullScreenClient::enterFullScreenForElement(WebPage* page
 }
 
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+void InjectedBundlePageFullScreenClient::updateImageSource(WebPage* page, FullScreenMediaDetails&& mediaDetails)
```
This doesn't belong in `InjectedBundlePageFullScreenClient` since you aren't actually interacting with the injected bundle. You can just send this message directly in `WebFullScreenManager::updateImageSource`.

aestes commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -470,6 +472,9 @@ void ImageLoader::notifyFinished(CachedResource& resource, const NetworkLoadMetr
         decode();
     loadEventSender().dispatchEventSoon(*this, eventNames().loadEvent);
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+    element().document().page()->chrome().client().updateImageSource(&element());
```
Why not pass `element()` as a reference instead of a pointer? It seems like WebFullscreenManager unconditionally dereferences it, so it's already not safe to pass null.

aestes replied:
Also:
1. I don't think it's safe to assume `page()` will be non-null.
2. You should call `protectedPage()` instead.
3. You should protect the element.

So you may want something like this:
```
if (RefPtr page = element().document().protectedPage())
    page->chrome().client().updateImageSource(*protectedElement());
```

aestes commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -241,6 +241,20 @@ void WebFullScreenManagerProxy::enterFullScreen(bool blocksReturnToFullscreenFro
+{
+    if (mediaDetails.imageHandle) {
+        auto sharedMemoryBuffer = SharedMemory::map(WTFMove(*mediaDetails.imageHandle), WebCore::SharedMemory::Protection::ReadOnly);
+        if (sharedMemoryBuffer)
```
```suggestion
        if (auto sharedMemoryBuffer = SharedMemory::map(WTFMove(*mediaDetails.imageHandle), WebCore::SharedMemory::Protection::ReadOnly))
```

aestes commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -221,51 +268,8 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
-
-        mediaDetails = getImageMediaDetails();
-    }
+    if (renderImage)
```
```suggestion
    if (CheckedPtr renderImage = dynamicDowncast<RenderImage>(element->renderer()))
```

aestes reviewed (approved):
r=me with comments

Commit:
[visionOS] Allow image updates when fullscreen is delegated to Quick Look
https://bugs.webkit.org/show_bug.cgi?id=281592
<rdar://136326912>

Reviewed by Andy Estes.

When an image source changes while in fullscreen with Quick look,
send that new image source over to Quick Look to update the image
being displayed.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::notifyFinished):
Listen for new image sources to finish loading.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::updateImageSource):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::updateImageSource):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::updateImageSource):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController updateImageSource]):
* Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::getImageMediaDetails):
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::updateImageSource):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::updateImageSource):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
Plumb the new image data to the UI Process if we currently have an
image open in fullscreen via Quick Look.

* Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift:
(item):
(windowOpenedContinuation):
(presentWindow):
(updateImage(_:)):
Wait until the Quick Look window is ready if it isn't already, then
update the Quick Look scene with the new image.

Canonical link: https://commits.webkit.org/285443@main
